In [ ]:
pip install opencv-python

# Task 2

For Task 2, since we are required to count the number of cars that go from the city’s downtown to the city centre, we first restrict the bounding to cars within that path and set up two points where cars will pass through, as not all cars will pass through the first point. 

#### Traffic_Laramie_1.mp4

In [3]:
import cv2
import threading
from prettytable import PrettyTable

table = PrettyTable()
initial_frame = None
counter = 0

# Create the video capture object
video=cv2.VideoCapture('Traffic_Laramie_1.mp4')

# Get frames per second
fps = video.get(cv2.CAP_PROP_FPS)
print("Frames per second: {0}".format(fps))
total = 0

# Start an infinite loop and keep reading frames until we press 'q'
while True:
    check, frame = video.read()

    # breaks once the end of the file is reached
    if frame is None:
        break
    # increment the total number of frames read
    total += 1
    gray_frame=cv2.cvtColor(frame,cv2.COLOR_BGR2GRAY)
    
    # The first captured frame is the baseline image
    if initial_frame is None:
        initial_frame = gray_frame
        print("Initial frame shape: ", initial_frame.shape)
        continue

    # The difference between the baseline and the gray frame
    delta_frame=cv2.absdiff(initial_frame,gray_frame)
    
    # The difference (the delta_frame) is converted into a binary image
    # If a particular pixel value is greater than a certain threshold (specified by us here as 150),
    # it will be assigned the value for White (255) else Black(0)
    threshold_frame=cv2.threshold(delta_frame,50,255, cv2.THRESH_BINARY)[1]
    blur_frame=cv2.GaussianBlur(threshold_frame,(25,25),0)
    (contours,_)=cv2.findContours(blur_frame,cv2.RETR_EXTERNAL,cv2.CHAIN_APPROX_SIMPLE)

    # Identify all the contours in our image.
    for c in contours:
        # Filter out any small contours
        if cv2.contourArea(c) < 6900:
            continue
        (x, y, w, h)=cv2.boundingRect(c)
        if y > 260:
            centroidX = int(x + w/2)
            centroidY = int(y + h/2)
            cv2.circle(frame, (centroidX, centroidY),1, (0,0,255), 2)
            cv2.rectangle(frame, (x, y), (x+w, y+h), (0,255,0), 1)
            if (centroidX <= 565 and centroidX >= 563) and (centroidY >= 350 and centroidY <= 470):
                counter = counter + 1
            elif (centroidX == 198) and (centroidY >= 350 and centroidY <= 430):
                counter = counter + 1
    cv2.putText(frame, "Number of cars: " + str(counter), (0, 185), cv2.FONT_HERSHEY_SIMPLEX, 1, (0,255,0), 2, cv2.LINE_AA)
            
    x1, y1 = 560, 350
    x2, y2 = 560, 470
    x3, y3 = 200, 350
    x4, y4 = 200, 430

    # Draw set up points
    line_thickness = 2
    cv2.line(frame, (x1, y1), (x2, y2), (0, 0, 255), thickness=line_thickness)
    cv2.line(frame, (x3, y3), (x4, y4), (0, 0, 255), thickness=line_thickness)
    cv2.imshow('Webcam', frame)
    
    # Stop the program by pressing 'q'    
    if cv2.waitKey(1) == ord('q'):
        break
print("Total number of cars: ", counter)
print("Total number of frames: ", total)
seconds = total/fps
minutes = seconds/60
carspermin = counter/minutes
print("Cars per minute: ", str(carspermin))

video.release()
cv2.destroyAllWindows()

table.field_names = ["Video", "Number of cars", "Cars per minute"]
table.add_row(["Traffic_Laramie_1.mp4", str(counter), str(carspermin)])
print(table)

Frames per second: 25.0
Initial frame shape:  (600, 1040)
Total number of cars:  6
Total number of frames:  4448
Cars per minute:  2.023381294964029
+-----------------------+----------------+-------------------+
|         Video         | Number of cars |  Cars per minute  |
+-----------------------+----------------+-------------------+
| Traffic_Laramie_1.mp4 |       6        | 2.023381294964029 |
+-----------------------+----------------+-------------------+


Cars per minute is calculated by first getting the frames per second and the total number of frames. Then, we get the total number of seconds by dividing the total number of frames by frames per second. After getting the seconds, we convert them into minutes by dividing by 60. Then, we get cars per minute by dividing the number of cars counted by the number of minutes. Thus, we will get (4448/25/60) minutes and 4 cars, which will derive 2.02 cars per minute. 

Now that we have generated the information needed, we do the same for the second video. However, revisions have to be made for the setup points to detect the vehicles passing through.

#### Traffic_Laramie_2.mp4

In [5]:
import cv2
import threading
from prettytable import PrettyTable

table = PrettyTable()
initial_frame = None
counter = 0

video=cv2.VideoCapture('Traffic_Laramie_2.mp4')
fps = video.get(cv2.CAP_PROP_FPS)
print("Frames per second: {0}".format(fps))
total = 0

while True:
    check, frame = video.read()
    if frame is None:
        break
    total += 1
    gray_frame=cv2.cvtColor(frame,cv2.COLOR_BGR2GRAY)

    if initial_frame is None:
        initial_frame = gray_frame
        print("Initial frame shape: ", initial_frame.shape)
        continue

    delta_frame=cv2.absdiff(initial_frame,gray_frame)
    threshold_frame=cv2.threshold(delta_frame,50,255, cv2.THRESH_BINARY)[1]
    blur_frame=cv2.GaussianBlur(threshold_frame,(25,25),0)
    (contours,_)=cv2.findContours(blur_frame,cv2.RETR_EXTERNAL,cv2.CHAIN_APPROX_SIMPLE)
    for c in contours:
        if cv2.contourArea(c) < 6900:
            continue
        (x, y, w, h)=cv2.boundingRect(c)
        if y > 260:
            centroidX = int(x + w/2)
            centroidY = int(y + h/2)
            cv2.circle(frame, (centroidX, centroidY),1, (0,0,255), 2)
            cv2.rectangle(frame, (x, y), (x+w, y+h), (0,255,0), 1)
            if (centroidX <= 406 and centroidX >= 400) and (centroidY >= 350 and centroidY <= 470):
                counter = counter + 1
            elif (centroidX == 198) and (centroidY >= 350 and centroidY <= 430):
                counter = counter + 1
    cv2.putText(frame, "Number of cars: " + str(counter), (0, 185), cv2.FONT_HERSHEY_SIMPLEX, 1, (0,255,0), 2, cv2.LINE_AA)
            
    x1, y1 = 400, 350
    x2, y2 = 400, 470
    x3, y3 = 200, 350
    x4, y4 = 200, 430

    line_thickness = 2
    cv2.line(frame, (x1, y1), (x2, y2), (0, 0, 255), thickness=line_thickness)
    cv2.line(frame, (x3, y3), (x4, y4), (0, 0, 255), thickness=line_thickness)
    cv2.imshow('Webcam', frame)
    
    if cv2.waitKey(1) == ord('q'):
        break

print("Total number of cars: ", counter)
print("Total number of frames: ", total)
seconds = total/fps
minutes = seconds/60
carspermin = counter/minutes
print("Cars per minute: ", str(carspermin))

video.release()
cv2.destroyAllWindows()

table.field_names = ["Video", "Number of cars", "Cars per minute"]
table.add_row(["Traffic_Laramie_2.mp4", str(counter), str(carspermin)])
print(table)

Frames per second: 25.0
Initial frame shape:  (600, 1040)
Total number of cars:  4
Total number of frames:  2642
Cars per minute:  2.271006813020439
+-----------------------+----------------+-------------------+
|         Video         | Number of cars |  Cars per minute  |
+-----------------------+----------------+-------------------+
| Traffic_Laramie_2.mp4 |       4        | 2.271006813020439 |
+-----------------------+----------------+-------------------+
